classes/form/management_component_form.php

```diff
@@ -0,0 +1,129 @@
+
+        $mform =& $this->_form;
+
+        // Set this variable to access correct db table.
+        $this->formtype = "component";
+
+        $mform->addElement('hidden', 'id');
+        $mform->setType('id', PARAM_INT);
+
+        $mform->addElement('text', 'name', get_string('componentname', 'tiny_elements'), ['size' => '255']);
+        $mform->setType('name', PARAM_TEXT);
+        $mform->addHelpButton('name', 'componentname', 'tiny_elements');
+        $mform->addRule('name', get_string('required'), 'required', null, 'client');
+        $mform->addRule('name', get_string('validclassname', 'tiny_elements'), 'regex', '/^[_a-zA-Z][_a-zA-Z0-9-]*$/', 'client');
+
+        $mform->addElement('text', 'displayname', get_string('displayname', 'tiny_elements'), ['size' => '255']);
+        $mform->setType('displayname', PARAM_TEXT);
+        $mform->addHelpButton('displayname', 'displayname', 'tiny_elements');
+
+        $mform->addElement('select', 'categoryname', get_string('category', 'tiny_elements'), $compcats);
+        $mform->setType('categoryname', PARAM_TEXT);
+        if (!empty($this->_ajaxformdata['categoryname'])) {
+            $mform->setDefault('categoryname', $this->_ajaxformdata['categoryname']);
+        }
+
+        $mform->addElement($this->codemirror_present() ? 'editor' : 'textarea', 'code', get_string('code', 'tiny_elements'));
+        $mform->setType('code', PARAM_RAW);
+        $mform->addHelpButton('code', 'code', 'tiny_elements');
+
+        $mform->addElement('textarea', 'text', get_string('text', 'tiny_elements'));
+        $mform->setType('text', PARAM_TEXT);
+
+        $mform->addElement(
+            'autocomplete',
+            'variants',
+            get_string('variants', 'tiny_elements'),
+            $variants,
+            ['multiple' => true, 'ajax' => 'tiny_elements/category_form_helper']
+        );
+        $mform->setType('variants', PARAM_TEXT);
+
+        $mform->addElement(
+            'autocomplete',
+            'flavors',
+            get_string('flavors', 'tiny_elements'),
+            $flavors,
+            ['multiple' => true, 'ajax' => 'tiny_elements/category_form_helper']
+        );
+        $mform->setType('flavors', PARAM_TEXT);
+
+        $mform->addElement('text', 'displayorder', get_string('displayorder', 'tiny_elements'));
+        $mform->setType('displayorder', PARAM_INT);
+
+        $mform->addElement($this->codemirror_present() ? 'editor' : 'textarea', 'css', get_string('css', 'tiny_elements'));
+        $mform->setType('css', PARAM_RAW);
+
+        $mform->addElement('url', 'iconurl', get_string('iconurl', 'tiny_elements'), ['size' => '255']);
+        $mform->setType('iconurl', PARAM_URL);
+
+        $mform->addElement(
+            'static',
+            'printurls',
+            '',
+            '<a href="#" onclick="window.open(\'printurls.php\', \'popup\', \'width=800,height=600\'); return false;">' .
```

Comment:
can this be moved over to a JS based modal? This will eventually trigger popup blocker I believe?

Reply:
Is now in a separate image picker